### ISSUE TITLE
Cannot Add Renderers within a Callback

### ISSUE THREAD

**Issue by guillochon** [bokeh/bokeh#4979]
In principle, it should be possible to add a renderer in a callback function in response to user input, this use case arises when the user wants to plot N datasets where N is variable and not known apriori. When doing this in the obvious way using the code below, the plots render very briefly before being blanked out, this appears to be a bug:

``` Python
from bokeh.layouts import column, row
from bokeh.models import Button, Range1d
from bokeh.models.widgets import TextInput
from bokeh.palettes import RdYlBu3
from bokeh.plotting import figure, curdoc

# create a plot and style its properties
tools = "pan,wheel_zoom,box_zoom,save,crosshair,reset,resize"
p = figure(title='Photometry Explorer', active_drag='box_zoom', y_range=Range1d(start=1, end=0),
           y_axis_label='Apparent Magnitude', tools=tools, plot_width=485, plot_height=485,
           toolbar_location='above', toolbar_sticky=False, x_range=Range1d(start=0, end=1))
p.xaxis.axis_label_text_font = 'futura'
p.yaxis.axis_label_text_font = 'futura'
p.xaxis.major_label_text_font = 'futura'
p.yaxis.major_label_text_font = 'futura'
p.xaxis.axis_label_text_font_size = '11pt'
p.yaxis.axis_label_text_font_size = '11pt'
p.xaxis.major_label_text_font_size = '8pt'
p.yaxis.major_label_text_font_size = '8pt'
p.title.align = 'center'
p.title.text_font_size = '16pt'
p.title.text_font = 'futura'
p.line(x=[], y=[], color="navy", line_width=3, line_join='round')

# add a button widget and configure with the call back
button = Button(label="Press Me")
namefield = TextInput(value="", title="Name(s):")
bandfield = TextInput(value="", title="Band(s):")

# add a text renderer to out plot (no data yet)

lineobs = []
circobs = []
new_datas = []
nds = []

def callback():
    new_datas = [{'x':[0.0, 1.0],'y':[0.0, 1.0]}]
    nxs = []
    nys = []
    for nd in new_datas:
        nxs.extend(nd['x'])
        nys.extend(nd['y'])
    x_buf = 0.1*(max(nxs) - min(nxs))
    y_buf = 0.1*(max(nys) - min(nys))

    p.x_range.start = min(nxs) - x_buf
    p.x_range.end = max(nxs) + x_buf
    p.y_range.start = max(nys) + y_buf
    p.y_range.end = min(nys) - y_buf

    for nd in new_datas:
        nds.append(nd)

        lineobs.append(p.line(x=[], y=[], color="navy", line_width=3, line_join='round'))
        circobs.append(p.circle(x=[], y=[], color="navy", line_width=3, line_join='round'))

        lineobs[-1].data_source.data = nds[-1]
        circobs[-1].data_source.data = nds[-1]

button.on_click(callback)

# put the button and plot in a layout and add to the document
curdoc().add_root(column(p, row(namefield, bandfield), button))
```

**Comment by bryevdv** [comment]
This seems to be triggering some kind of infinite ping-pong in the code that detects changes and invalidates/recomputes all the models. This happens after a single button press and does not stop:

<img width="736" alt="screen shot 2016-08-17 at 3 24 20 pm" src="https://cloud.githubusercontent.com/assets/1078448/17752637/5f9f1a9e-6492-11e6-9fa9-b8c070884a84.png">

**Comment by bryevdv** [comment]
@guillochon I believe this was fixed by some fixes for server sync in the `0.12.5` release  cycle:

<img width="586" alt="screen shot 2017-03-24 at 11 52 37" src="https://cloud.githubusercontent.com/assets/1078448/24304728/74822bac-1088-11e7-9b39-8c769f3bd765.png">

If you can still reproduce problems with `0.12.5` please feel free to e-open
